File: bank_connector_erpnext/erpnext___bank_connector/override/payment_order.py
import frappe, json
from erpnext.accounts.doctype.payment_order.payment_order import PaymentOrder
from bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_order import make_payment_entries

class CustomPaymentOrder(PaymentOrder):
	def validate(self):
		self.validate_summary()

	def validate_summary(self):
		if len(self.summary) <= 0:
			frappe.throw("Please validate the summary")
		
		default_mode_of_transfer = None
		if self.default_mode_of_transfer:
			default_mode_of_transfer = frappe.get_doc("Mode of Transfer", self.default_mode_of_transfer)

		for payment in self.summary:
			if payment.mode_of_transfer:
				mode_of_transfer = frappe.get_doc("Mode of Transfer", payment.mode_of_transfer)
			else:
				if not default_mode_of_transfer:
					frappe.throw("Define a specific mode of transfer or a default one")
				mode_of_transfer = default_mode_of_transfer
				payment.mode_of_transfer = default_mode_of_transfer.mode

			if payment.amount < mode_of_transfer.minimum_limit or payment.amount > mode_of_transfer.maximum_limit:
				frappe.throw(f"Mode of Transfer not suitable for {payment.party} for {payment.amount}. {mode_of_transfer.mode}: {mode_of_transfer.minimum_limit}-{mode_of_transfer.maximum_limit}")

		summary_total = 0
		references_total = 0
		for ref in self.references:
			references_total += ref.amount
		
		for sum in self.summary:
			summary_total += sum.amount

		if summary_total != references_total:
			frappe.throw("Summary isn't matching the references")

	def on_submit(self):
		make_payment_entries(self.name)
		frappe.db.set_value("Payment Order", self.name, "status", "Pending")

		for ref in self.references:
			if hasattr(ref, "payment_request"):
				frappe.db.set_value("Payment Request", ref.payment_request, "status", "Payment Ordered")

	def on_update_after_submit(self):
		frappe.throw("You cannot modify a payment order")
		return


	def before_cancel(self):
		frappe.throw("You cannot cancel a payment order")
		return
	
	def on_trash(self):
		if self.docstatus == 1:
			frappe.throw("You cannot delete a payment order")
			return


@frappe.whitelist()
def get_party_summary(references, company_bank_account):
	references = json.loads(references)
	if not len(references) or not company_bank_account:
		return

	# Considering the following dimensions to group payments
	# (party_type, party, bank_account, account, cost_center, project)

	summary = {}
	for ref in references:
		ref = frappe._dict(ref)
		if (ref.party_type, ref.party, ref.bank_account, ref.account, ref.cost_center, ref.project, ref.tax_withholding_category, ref.reference_doctype) in summary:
			summary[(ref.party_type, ref.party, ref.bank_account, ref.account, ref.cost_center, ref.project, ref.tax_withholding_category, ref.reference_doctype)] += ref.amount
		else:
			summary[(ref.party_type, ref.party, ref.bank_account, ref.account, ref.cost_center, ref.project, ref.tax_withholding_category, ref.reference_doctype)] = ref.amount

	result = []
	for k, v in summary.items():
		party_type, party, bank_account, account, cost_center, project, tax_withholding_category, reference_doctype = k
		summary_line_item = {}
		summary_line_item["party_type"] = party_type
		summary_line_item["party"] = party
		summary_line_item["bank_account"] = bank_account
		summary_line_item["account"] = account
		summary_line_item["cost_center"] = cost_center
		summary_line_item["project"] = project
		summary_line_item["tax_withholding_category"] = tax_withholding_category
		summary_line_item["reference_doctype"] = reference_doctype
		summary_line_item["amount"] = v
		result.append(summary_line_item)
	
	for row in result:
		party_bank = frappe.db.get_value("Bank Account", row["bank_account"], "bank")
		company_bank = frappe.db.get_value("Bank Account", company_bank_account, "bank")
		row["mode_of_transfer"] = None
		if party_bank == company_bank:
			mode_of_transfer = frappe.db.get_value("Mode of Transfer", {"is_bank_specific": 1, "bank": party_bank})
			if mode_of_transfer:
				row["mode_of_transfer"] = mode_of_transfer
		else:
			mot = frappe.db.get_value("Mode of Transfer", {
				"minimum_limit": ["<=", row["amount"]], 
				"maximum_limit": [">", row["amount"]],
				"is_bank_specific": 0
				}, 
				order_by = "priority asc")
			if mot:
				row["mode_of_transfer"] = mot
	
	return result

File: bank_connector_erpnext/public/js/payment_order.js
frappe.ui.form.on('Payment Order', {
	onload(frm) {
		frm.set_df_property("payment_order_type", "options", [""].concat(["Payment Request", "Payment Entry", "Purchase Invoice"]));
		frm.refresh_field("payment_order_type");
	},
	refresh(frm) {
		frm.set_df_property('summary', 'cannot_delete_rows', true);
		frm.set_df_property('summary', 'cannot_add_rows', true);
		frm.remove_custom_button("Payment Entry", "Get Payments from");
		frm.remove_custom_button("Payment Request", "Get Payments from");
		frm.set_df_property("payment_order_type", "options", [""].concat(["Payment Request", "Payment Entry", "Purchase Invoice"]));
		frm.refresh_field("payment_order_type");
		if (frm.doc.docstatus == 0) {
			frm.add_custom_button(__('Payment Request'), function() {
				frm.trigger("remove_row_if_empty");
				erpnext.utils.map_current_doc({
					method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_request.make_payment_order",
					source_doctype: "Payment Request",
					target: frm,
					setters: {
						party: frm.doc.supplier || "",
						grand_total: "",
					},
					get_query_filters: {
						docstatus: 1,
						status: ["in", ["Initiated"]],
						mode_of_payment: "Wire Transfer",
						transaction_date : ["<=", frm.doc.posting_date],
						company: frm.doc.company
					}
				});
			}, __("Get from"));
			frm.add_custom_button(__('Payment Entry'), function() {
				frm.trigger("remove_row_if_empty");
				erpnext.utils.map_current_doc({
					method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_entry.make_payment_order",
					source_doctype: "Payment Entry",
					target: frm,
					setters: {
						party: frm.doc.supplier || "",
						paid_amount:"",
					},
					get_query: function() {
						var filters = {
							company: frm.doc.company,
							posting_date : ["<=", frm.doc.posting_date],
							payment_type:"Pay"
						};
						return {
							query: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_entry.fetch_unprocessed_payment_entries",
							filters: filters
						};
					}
				});
			}, __("Get from"));

			// frm.add_custom_button(__('Purchase Invoice'), function() {
			// 	frm.trigger("remove_row_if_empty");
			// 	erpnext.utils.map_current_doc({
			// 		method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.purchase_invoice.make_payment_order",
			// 		source_doctype: "Purchase Invoice",
			// 		target: frm,
			// 		setters: {
			// 			supplier: "",
			// 			outstanding_amount: "",
			// 			status: ""
			// 		},
			// 		get_query_filters: {
			// 			docstatus: 1,
			// 			on_hold: ["!=", 1],
			// 			due_date : ["<=", frm.doc.posting_date],
			// 			outstanding_amount: [">", 0]
			// 		}
			// 	});
			// }, __("Get from"));
		};
		if (frm.doc.docstatus===1 && frm.doc.payment_order_type==='Payment Request') {
			frm.remove_custom_button(__('Create Payment Entries'));
		}
		if (frm.doc.status == "Pending" && frm.doc.docstatus == 1) {
			if (frm.has_perm('write') && 'summary' in frm.doc) {
				var uninitiated_payments = 0;
				for(var i = 0; i < frm.doc.summary.length; i++) {
					if (!frm.doc.summary[i].payment_initiated) {
						uninitiated_payments += 1
					}
				}
				if (uninitiated_payments > 0) {
					frm.add_custom_button(__('Initiate Payment'), function() {
						// check has_otp in bank connector
						frappe.db.get_value("Bank Connector", {"bank_account": cur_frm.doc.company_bank_account}, 'has_otp', (r) => {
							frappe.call({
								method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_order.call_otp_sender",
								args: {
									docname: cur_frm.doc.name,
									bank_account: cur_frm.doc.company_bank_account,
									total_amount: cur_frm.doc.total,
								},
								callback: function(e) {
									frappe.msgprint(e.message);
								}
							})
							if (r['has_otp']) {
								frappe.prompt({
									label: 'Enter OTP',
									fieldname: 'otp',
									fieldtype: 'Data'
								}, (values) => {
									let cur_otp = values.otp;
									frappe.call({
										method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_order.make_bank_payment",
										args: {
											docname: cur_frm.doc.name,
											bank_account: cur_frm.doc.company_bank_account,
											otp: cur_otp
										},
										callback: function(r) {
											if (r.message) {
												frappe.msgprint(r.message)
											}
										}
									})
								});
							}
						}); 	
					});
				}
			}
		}

		if ((frm.doc.status == "Pending" || frm.doc.status == "Initiated") && frm.doc.docstatus == 1) {
			if (frm.has_perm('write') && 'summary' in frm.doc) {
				var pending_status_check = 0
				for (var j = 0; j < frm.doc.summary.length; j++) {
					if(frm.doc.summary[j].payment_status == "Initiated") {
						pending_status_check += 1
					}
				}

				if (pending_status_check > 0) {
					frm.remove_custom_button(__('Initiate Payment'))
					frm.add_custom_button(__('Get Status'), function() {
						frappe.call({
							method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_order.get_payment_status",
							freeze: 1,
							args: {
								docname: frm.doc.name,
								bank_account: frm.doc.company_bank_account,
							},
							callback: function(r) {
								if(r.message) {
									frappe.msgprint(r.message)
								}
								frm.reload_doc();
							}
						});
					});
				}
			}
		}


	},
	remove_button: function(frm) {
		// remove custom button of order type that is not imported
		let label = ["Payment Request", "Purchase Invoice"];

		if (frm.doc.references.length > 0 && frm.doc.payment_order_type) {
			label = label.reduce(x => {
				x!= frm.doc.payment_order_type;
				return x;
			});
			frm.remove_custom_button(label, "Get from");
		}
	},
	get_summary: function(frm) {
		if (frm.doc.docstatus > 0) {
			frappe.msgprint("Not allowed to change post submission");
			return
		}
		frappe.call({
			method: "bank_connector_erpnext.erpnext___bank_connector.override.payment_order.get_party_summary",
			args: {
				references: frm.doc.references,
				company_bank_account: frm.doc.company_bank_account
			},
			freeze: true,
			callback: function(r) {
				if(r.message) {
					let summary_data = r.message
					frm.clear_table("summary");
					var doc_total = 0
					for (var i = 0; i < summary_data.length; i++) {
						doc_total += summary_data[i].amount
						let row = frm.add_child("summary");
						row.party_type = summary_data[i].party_type;
						row.party = summary_data[i].party;
						row.amount = summary_data[i].amount;
						row.bank_account = summary_data[i].bank_account;
						row.account = summary_data[i].account;
						row.mode_of_transfer = summary_data[i].mode_of_transfer;
						row.cost_center = summary_data[i].cost_center;
						row.project = summary_data[i].project;
						row.tax_withholding_category = summary_data[i].tax_withholding_category;
						row.reference_doctype = summary_data[i].reference_doctype;
					}
					frm.refresh_field("summary");
					frm.doc.total = doc_total;
					frm.refresh_fields();
				}
			}
		});
	},
	update_status: function(frm) {
		if (frm.doc.docstatus != 1) {
			frappe.msgprint("Updating status is not allowed without submission");
			return
		}

		if (!frm.doc.approval_status) {
			frappe.msgprint("Updating status is not allowed without value");
			return
		}

		var selected_rows = frm.get_selected()
		if (!Object.keys(selected_rows).length || !"summary" in selected_rows){
			frappe.msgprint("No rows are selected");
			return
		}

		frappe.call({
			method: "bank_connector_erpnext.erpnext___bank_connector.doc_events.payment_order.modify_approval_status",
			args: {
				items: selected_rows.summary,
				approval_status: frm.doc.approval_status,
			},
			callback: function(r) {
				if(r.message) {
					var updated_count = 0
					for (var line_item in r.message) {
						if (r.message[line_item].status) {
							frappe.model.set_value("Payment Order Summary", line_item, "approval_status", r.message[line_item].message);
							updated_count += 1
						} else {
							frappe.msgprint(r.message[line_item].message)
						}
					}
					frappe.msgprint(updated_count + " record(s) updated.")
				}
				frm.dirty();
				frm.refresh_fields();
			}
		});
	}

});

frappe.ui.form.on('Payment Order Summary', {
	setup: function(frm) {
		frm.set_query("party_type", function() {
			return {
				query: "erpnext.setup.doctype.party_type.party_type.get_party_type",
			};
		});
	}
})